Stripe/Payments/STPPaymentManager.m

```diff
@@ -0,0 +1,406 @@
+                                                                                 code:STPPaymentManagerRequiresPaymentMethodErrorCode
+                                                                             userInfo:nil]);
+                break;
+            case STPPaymentIntentStatusRequiresConfirmation:
+                completion(STPPaymentManagerActionStatusSucceeded, paymentIntent, nil);
+                break;
+            case STPPaymentIntentStatusRequiresAction:
+                if (attemptAuthentication) {
+                    [self _handleAuthenticationForCurrentAction];
+                } else {
+                    // If we get here, the user exited from the redirect before the
+                    // payment intent was updated. Consider it a cancel
+                    completion(STPPaymentManagerActionStatusCanceled, paymentIntent, nil);
+                }
+                break;
+            case STPPaymentIntentStatusProcessing:
+                completion(STPPaymentManagerActionStatusSucceeded, paymentIntent, nil);
+                break;
+            case STPPaymentIntentStatusSucceeded:
+                completion(STPPaymentManagerActionStatusSucceeded, paymentIntent, nil);
+                break;
+            case STPPaymentIntentStatusRequiresCapture:
+                completion(STPPaymentManagerActionStatusSucceeded, paymentIntent, nil);
+                break;
+            case STPPaymentIntentStatusCanceled:
+                completion(STPPaymentManagerActionStatusCanceled, paymentIntent, nil);
+                break;
+        }
+    } else {
+        completion(STPPaymentManagerActionStatusFailed, nil, [NSError errorWithDomain:STPPaymentManagerErrorDomain
+                                                                              code:STPPaymentManagerInternalErrorCode
+                                                                          userInfo:nil]);
+    }
+}
+
+- (void)_handleAuthenticationForCurrentAction {
+    STPPaymentIntent *paymentIntent = _currentAction.paymentIntent;
+    STPPaymentIntentAction *authenticationAction = paymentIntent.nextAction;
+    STPPaymentManagerActionCompletionBlock completion = _currentAction.completion;
+    if (completion == nil) {
+        return;
+    }
+
+    switch (authenticationAction.type) {
+
+        case STPPaymentIntentActionTypeUnknown:
+            completion(STPPaymentManagerActionStatusFailed, paymentIntent, [NSError errorWithDomain:STPPaymentManagerErrorDomain
+                                                                             code:STPPaymentManagerUnsupportedAuthenticationErrorCode
+                                                                         userInfo:nil]);
+            break;
+        case STPPaymentIntentActionTypeRedirectToURL: {
+            NSURL *url = authenticationAction.redirectToURL.url;
+
+            [[STPURLCallbackHandler shared] registerListener:self forURL:authenticationAction.redirectToURL.returnURL];
```

Comment:
Might need to check `authenticationAction.redirectToURL.returnURL != nil`

Reply:
👍